### PR TITLE
Feature/payment integration #20

### PR DESCRIPTION
**Pull request by DioneMartin**
### Description
This PR implements a robust, multi-country tax payment orchestration system within the Check-in flow. It enables regionalized payment methods (3DS, WIRE, GOV) for US, MX, and CA, ensuring all legal and technical requirements are met before a boarding pass is issued.

Motivation and Context:
To support international expansion, the system now validates country-specific payment rules and integrates with an external Payment Service. This change ensures that mandatory data (like CURP for Mexico or Card Numbers for 3DS) is captured and validated through an asynchronous tokenization process.

Fixed Issue: #20 

Dependencies:

external.payment-service.url environment variable.

Updated Wiremock server for integration test simulations.

### Technical Details
-**Modules Affected:** service, http, integration, and docker.

-**Key Changes:**

-   Flow Orchestration: Added PaymentMethodStep, PaymentValidationStep, and TransactionFinalizationStep to the GeneralContinueFlow. This ensures a sequential validation of availability, data completeness, and final status check.

-   Adapter Architecture: Created a decoupled payment system using PaymentFacade and PaymentAdapterFactory. Specific implementations (GovPaymentAdapter, ThreeDsPaymentAdapter, WireTransferPaymentAdapter) handle the unique validation logic for each method.

-   External Communication: Implemented PaymentIntegrationAdapter to bridge the service layer with external HTTP clients (PaymentClient, PaymentStatusClient) using Spring’s declarative interface clients.

-   Configuration-Driven Logic: Used application.properties and Spring Environment to manage allowed payment methods per country, allowing for easy updates without code changes.

-   Testing Infrastructure:

-   Unit Tests: Added comprehensive JUnit 5/Mockito tests for all new Steps and Adapters, reaching high coverage of edge cases (e.g., missing CURP or unauthorized country-method combinations).

-   Integration Tests: Implemented PaymentMethodFlowTest in Spock/Wiremock to simulate end-to-end check-in scenarios, including "bypass" logic and error handling for rejected payment methods.

-   Mocking: Added 4 new Wiremock mappings to simulate various payment token scenarios and successful status responses.

### CODE REVIEW

**Comment by CarlosHPlata** [comment]
Can you move all  code from payment in service (`service/src/main/java/com/aterrizar/service/payment`), to the http folder, create a package in `http/src/main/java/com/aterrizar/http/external/gateway` just like the `aviator` one, remember the folder names and package, should reflect intent of the code, not what the code does, so try to avoid names like `adapter` for packages, and for classes it can be used as a suffix like you did with `PaymentAdapter`

File: http/src/main/java/com/aterrizar/http/external/gateway/PaymentStatusClient.java
Comment (CarlosHPlata) [inline http/src/main/java/com/aterrizar/http/external/gateway/PaymentStatusClient.java:1]: Why does this class exists if it's mapped on the `PaymentClient`?

File: service/src/main/java/com/aterrizar/service/payment/factory/PaymentAdapterFactory.java
Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/payment/factory/PaymentAdapterFactory.java:1]: No need for an interface of the factory.

File: service/src/main/java/com/aterrizar/service/payment/facade/PaymentFacade.java
Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/payment/facade/PaymentFacade.java:1]: This class should be moved to `service/src/main/java/com/aterrizar/service/external` call this `PaymentGateway`

File: service/src/main/java/com/aterrizar/service/payment/facade/impl/PaymentFacadeImpl.java
Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/payment/facade/impl/PaymentFacadeImpl.java:1]: This is the main point of access from the `http` once moved, call it `PaymentFacade` implements `PaymentGateway`

File: service/src/main/java/com/aterrizar/service/checkin/steps/PaymentMethodStep.java
Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/checkin/steps/PaymentMethodStep.java:1]: I was not expecting the property to be handled on a step, please create a configuration class just like `DigitalVisaConfig` is in `http/src/main/java/com/aterrizar/http/config/feature/DigitalVisaConfig.java`

Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/checkin/steps/PaymentMethodStep.java:1]: Also this step should be called `ValidPaymentMethodCheck` or something.

File: service/src/main/java/com/aterrizar/service/checkin/steps/PaymentValidationStep.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import com.aterrizar.service.payment.facade.PaymentFacade;
+import com.aterrizar.service.payment.model.PaymentRequestDto;
+
+@Component
+public class PaymentValidationStep implements Step {
+
+    private final PaymentFacade paymentFacade;
+    private final Environment environment;
+
+    public PaymentValidationStep(PaymentFacade paymentFacade, Environment environment) {
```
Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/checkin/steps/PaymentValidationStep.java]: use `@RequiredArgsConstructor` instead of putting all vars in the constructor.

File: service/src/main/java/com/aterrizar/service/checkin/steps/PaymentValidationStep.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        Map<RequiredField, String> fields = context.checkinRequest().providedFields();
+        return fields != null
+                && fields.containsKey(RequiredField.PAYMENT_METHOD)
+                && (context.session().sessionData() == null
+                        || context.session().sessionData().paymentToken() == null);
+    }
+
+    @Override
+    public StepResult onExecute(Context context) {
+
```
Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/checkin/steps/PaymentValidationStep.java]: extra line

File: service/src/main/java/com/aterrizar/service/checkin/steps/PaymentValidationStep.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        String countryCode = context.countryCode().name();
+        String allowedMethodsProperty =
+                environment.getProperty("feature.tax.payments." + countryCode, "");
```
Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/checkin/steps/PaymentValidationStep.java]: same comment as `PaymentMethodStep` you should not consume the properties directly here

File: service/src/main/java/com/aterrizar/service/checkin/steps/PaymentValidationStep.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        if (missingField != null) {
+            return StepResult.terminal(context.withRequiredField(missingField));
+        }
+
+        String paymentToken = paymentFacade.executePayment(requestDto);
+        return StepResult.success(context.withSessionData(data -> data.paymentToken(paymentToken)));
+    }
+
+    private RequiredField checkMissingSpecificField(PaymentRequestDto dto) {
+        return switch (dto.paymentMethod().toUpperCase()) {
```
Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/checkin/steps/PaymentValidationStep.java]: The step is currently leaking implementation details about payment methods. This violates the purpose of our Adapter (step should be agnostic about which specific methods exists). Let's move this selection logic to the Facade so that the step only sees a single, generic payment interface.

File: service/src/main/java/com/aterrizar/service/checkin/steps/TransactionFinalizationStep.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                && context.session().sessionData().paymentToken() != null;
+    }
+
+    @Override
+    public StepResult onExecute(Context context) {
+        String token = context.session().sessionData().paymentToken();
+
+        // Using the Port to comply with the Mock Behavior: GET /payment-service/v1/status/{token}
+        String status = externalPaymentPort.getPaymentStatus(token);
+        boolean isSuccess = "SUCCESS".equalsIgnoreCase(status);
```
Comment (CarlosHPlata) [inline service/src/main/java/com/aterrizar/service/checkin/steps/TransactionFinalizationStep.java]: magic string, move it to a constant

File: integration/src/test/groovy/com/aterrizar/integration/test/feature/PaymentMethodFlowTest.groovy [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Unroll
+    def "should complete flow for #method"() {
+        setup:
+        def checkin = Checkin.create()
+
+        when: "Complete flow with all required data"
+        def session = checkin.initSession(country)
+        def response = session.fillUserInput([
```
Comment (CarlosHPlata) [inline integration/src/test/groovy/com/aterrizar/integration/test/feature/PaymentMethodFlowTest.groovy:71]: Can we divide the steps, first test that the field is requested, then send the field filled and see if pass.

**Review by CarlosHPlata — CHANGES_REQUESTED** [review]

**Comment by VictorFloresJ** [comment]
Hi @CarlosHPlata ,

We have applied all the requested changes:

- Architecture: Moved payment logic to the http module and implemented PaymentGateway in service to ensure proper dependency inversion.
- Configuration: Replaced Environment with PaymentConfig (using ConfigurationProperties) to decouple Steps from infrastructure.
- Clean Code: Moved field-specific validation to the Facade, removed the redundant factory interface, and replaced magic strings with constants in PaymentGateway.
- Testing Status: The Groovy integration tests are still being adjusted to match the requirements, but we wanted to push the core refactor for your review

The project builds successfully with ./gradlew build. We hope you like it 
With love,
KISS

**Comment by CarlosHPlata** [comment]
Bravo, there's still some issues like lack of unit test, and some nit pick, but the overall solution is what I was expecting, I can't be more happy of how you guys handle it. Really well done
